relay-server/src/actors/processor.rs

```diff
@@ -1267,6 +1268,95 @@ impl EnvelopeProcessorService {
+                            .expect("This is pre-validated and should not fail.");
+
+                    // We need to use this value twice. Once for pre-allocating memory and once
+                    // for serialization.
+                    let replay_event = replay_event_item.payload();
+
+                    // Pre-allocate capacity in the buffer. The buffer will use the full size of
+                    // each of the three keys' values. As well as the size of the keys themselves
+                    // (52 bytes) plus some overhead of unknown size. 64 was used as a safe guess
+                    // for the combined sizes of the keys and msgpack bytes.
+                    //
+                    // Pre-allocating capacity makes pushing into the buffer essintially free.
+                    let capacity = replay_event.len() + headers.len() + body.len() + 64;
+                    let mut buffer: Vec<u8> = Vec::with_capacity(capacity);
+
+                    // Msgpack serialization step. We take all of the data we've extracted and
+                    // serialize it into a single payload.
+                    //
+                    // TODO: For now I've inlined the Payload definition. We can argue over its
+                    // best location in code review.
+                    #[derive(Debug, Serialize)]
+                    struct Payload {
+                        replay_event: Vec<u8>,
+                        replay_recording_headers: Vec<u8>,
+                        replay_recording: Vec<u8>,
+                    }
+
+                    let payload = Payload {
+                        replay_event: replay_event.into(),
+                        replay_recording_headers: headers,
+                        replay_recording: body,
+                    };
+
+                    if let Err(e) = rmp_serde::encode::write(&mut buffer, &payload) {
+                        // Log so we can monitor failures.
+                        relay_log::error!("replay envelope merge failure: {e}");
+
+                        // If the merging process fails we need to re-append the old envelopes
+                        // back into the stream so they can be processed in the version 0
+                        // style.
+                        envelope.add_item(replay_event_item);
+                        envelope.add_item(replay_recording_item);
+                    } else {
+                        // If the operation was successful we can create our new combined envelope
+                        // items and append it to the envelope stream.
+                        let mut combined_item =
+                            Item::new(ItemType::CombinedReplayEventAndRecording);
+                        combined_item.set_payload(ContentType::MsgPack, buffer);
+                        envelope.add_item(combined_item);
+                    }
+                } else {
+                    envelope.add_item(replay_event_item)
+                }
```

Comment:
Instead of taking the `ReplayEvent` and putting it back when there is no recording, can we take the recording first, and only remove the `ReplayEvent` if it exists?

Reply:
If the replay-event doesn't exist then we would have to add the recording back as an envelope item.  There's no benefit to this swap.